ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1295,4 +998,533 @@ private void endCompaction() {
+      //                     | Delete(f)         | false                                        |
+      //                     | Undelete(f)       | c==f&&!isExpired(Uc)?true:false              |
+      // ----------------------------------------------------------------------------------------
+      List<IndexEntry> validEntries = new ArrayList<>();
+      Iterator<IndexEntry> iterator = indexSegment.getIterator();
+      StoreKey lastKey = null;
+      IndexValue lastFinalState = null;
+      while (iterator.hasNext()) {
+        IndexEntry entry = iterator.next();
+        StoreKey currentKey = entry.getKey();
+        IndexValue currentValue = entry.getValue();
+        IndexValue currentFinalState;
+
+        if (lastKey == null) {
+          // save a token for restart (the key gets ignored but is required to be non null for construction)
+          StoreFindToken safeToken =
+              new StoreFindToken(currentKey, indexSegment.getStartOffset(), sessionId, incarnationId);
+          compactionLog.setSafeToken(safeToken);
+          logger.debug("Set safe token for compaction in {} to {}", storeId, safeToken);
+        }
+        if (currentKey.equals(lastKey)) {
+          currentFinalState = lastFinalState;
+        } else {
+          lastKey = currentKey;
+          lastFinalState = currentFinalState = srcIndex.findKey(currentKey);
+        }
+
+        if (currentValue.isUndelete()) {
+          if (currentFinalState.isPut()) {
+            throw new IllegalStateException(
+                "Undelete's final state can't be put for key" + currentKey + " in store " + dataDir);
+          }
+          if (currentFinalState.isUndelete() && currentFinalState.getLifeVersion() == currentValue.getLifeVersion()
+              && !srcIndex.isExpired(currentValue)) {
+            validEntries.add(entry);
+          }
+        } else if (currentValue.isDelete()) {
+          if (currentFinalState.isPut()) {
+            throw new IllegalStateException(
+                "Delete's final state can't be put for key" + currentKey + " in store " + dataDir);
+          }
+          if (currentFinalState.isDelete() && currentFinalState.getLifeVersion() == currentValue.getLifeVersion()) {
+            validEntries.add(entry);
+          }
+        } else if (currentValue.isTtlUpdate()) {
+          if (currentFinalState.isPut()) {
+            // If isPut returns true, when the final state doesn't carry ttl_update flag, this is wrong
+            throw new IllegalStateException(
+                "TtlUpdate's final state can't be put for key" + currentKey + " in store " + dataDir);
+          }
+          // This is a TTL_UPDATE record, then the blob can't be expired. Only check if it's deleted or not.
+          if (currentFinalState.isDelete()) {
+            if (currentFinalState.getOperationTimeInMs() >= compactionLog.getCompactionDetails().getReferenceTimeMs()) {
```

Comment:
Can you explain the intention of `currentFinalState.getOperationTimeInMs() >= compactionLog.getCompactionDetails().getReferenceTimeMs()`?
Also looks like if `CurrentFinalState.isDelete()`, the if/else if can combined together.

Reply:
`compactionLog.getCompactionDetails().getReferenceTimeMs()` is set to be the current time - retention duration when starting compaction. So if the retention duration is 7 days, then the reference time is now - 7d. And this if statement means if the delete's operation is later than now - 7d (e.g. now - 1d), the delete and ttl update should not be compacted.

I have two if statement under this `if (currentFinalState.isDelete())`, it's just easier to read in this way.